[IOSP-607] Update PullAssigners document to remove reference to Pull Panda and explain GitHub's code review assignment (#359)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/PullAssigners.md

## Introduction

The aim of this document is to explain how to configure GitHub's automatic code review assignment in our Babylon repos in order for it to assign people to PRs automatically.

- We will use the `CODEOWNERS` file to make GitHub always assign a GitHub team to all our Pull Requests.

> - GitHub will then pick N reviewers from that team ([see below how to configure that number N](#enabling-github-s-code-review-assignment)) and unassign the team itself.

Reviewer: ```suggestion
- GitHub will then pick N reviewers from that team ([see below how to configure that number N](#enabling-githubs-code-review-assignment)) and unassign the team itself.
```

> - GitHub will then pick N reviewers from that team and unassign the team itself.

Reviewer: My immediate question when reading this was "Where is that N configured?". Then continuing reading on I saw the section further below explaining it.
So might be a good idea to link to the instructions below from here

```suggestion
- GitHub will then pick N reviewers from that team ([see below how to configure that number N](#enabling-github-s-code-review-assignment)) and unassign the team itself.
```

Author: Good suggestion 👍 The link doesn't seem to be working in the file viewer so I'll check it when it's merged.

Reviewer: Checked the anchor via Google Chrome inspected while in file viewer, the proper link should be this instead: `#enabling-githubs-code-review-assignment`

- Optionally, we could also require each PR of a repository to always have an approval from a subset of people (the ones owning the responsibility of the code) – like we did for our bots repository owned by the Developer Experience squad.

## Configure GitHub Teams

> ### A team to list the reviewers to pick from

Reviewer: Since this level-2 section doesn't have more than one level-3 subsection anymore (now that you deleted all the other subsections that are not needed anymore), this title hierarchy is a bit redundant

```suggestion
## Configure a GitHub Team containing the reviewers to pick from
```

> 1. First you need a GitHub team containing the list of all the people you want GitHub to pick reviewers from.

Reviewer 2: ```suggestion
1. First, you need a GitHub team containing the list of all the people you want GitHub to pick reviewers from.
```

   * If you need a different list to pick your reviewers from, you'd need to create a team via [this page on GitHub](https://github.com/orgs/babylonhealth/new-team), and once it's created, go to the team's page and add Members to that new team ([e.g. here for `iOS-Devs`](https://github.com/orgs/babylonhealth/teams/iOS-Devs/members))

2. Next, **make sure that this GitHub team –containing the people you want as reviewers– has access to your repository**, via the "Repositories" tab in the team's page ([e.g. here for `iOS-Devs`](https://github.com/orgs/babylonhealth/teams/iOS-Devs/repositories))

## The `CODEOWNERS` file

Now we need to ensure that the right team (e.g. `iOS-Devs`) is automatically assigned as reviewer on all new PRs.

For this, you just need to create a `.github/CODEOWNERS` file in your repo with the following content (if you have a `CODEOWNERS` file already at the root of your repository, we advise to move it inside a `.github/` directory to clean up your repo root).

```
# Global rule for the whole codebase.
#  - The empty `iOS-PullAssigner` team will act as a proxy for the PullAssigner bot.

> #    When a new PR arrives, PullAssigner will be triggered, which will then pick members from the iOS-Admin team

Reviewer: This comment in the `CODEOWNERS` file should be removed as it mentions PullAssigner (and we don't use the concept of a proxy team anymore with GitHub method anyway)

You might want to update the comment to something like this instead
```suggestion
# Global rule for the whole codebase.
# GitHub's Code Reviewer assignment feature will only pick N people from that list
#   to assign them as reviewers. To configure this, go to your Team's page
#   (https://github.com/orgs/babylonhealth/teams/<yourteam>) then to
#   Settings > Code review assignment
```

Author: Oops, missed this one (GitHub's search is terrible). Thanks for spotting!

* @babylonhealth/iOS-Devs
```

## Enabling GitHub's code review assignment

Once the team and `CODEOWNERS` file has been created, we can go ahead and enable GitHub's reviewer assignment feature.

1. Head to the organisation's [home page](https://github.com/babylonhealth).
2. Go to the "Teams" tab and search for your team ([e.g. here for `iOS-Devs`](https://github.com/orgs/babylonhealth/teams/ios-devs)).
3. Go to the "Settings" page (to access / edit your team's settings you'll need to have the `Maintainer` team role).
4. In "Settings" go to "Code review assignment".
5. Tick "Enable auto assignment" and configure the settings as desired. 

> **Note:** You'll probably want to tick "If assigning team members, don't notify the entire team" to prevent the entire team getting notifications for every PR, even if they haven't been individually assigned to it.

Reviewer 2: ```suggestion
**Note:** You'll probably want to tick "If assigning team members, don't notify the entire team" to prevent the entire team from getting notifications for every PR, even if they haven't been individually assigned to it.
```

Please see [GitHub's documentation](https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/managing-code-review-assignment-for-your-team) for more in-depth details around configuring these settings.

#### In summary

Now that we have a team of reviewers, your project has a `CODEOWNERS` file and GitHub code review assignment is enabled, the following will happen when you create a PR:

1. The team defined in `CODEOWNERS` will be automatically assigned as a reviewer.
2. GitHub will instantly pick N reviewers and assign them to the pull request (number of reviewers is configured in the "Code review assignment" settings).
3. The team will be unassigned.

Congratulations, you now have reviewers for your PR and you didn't need to lift a finger.

#### Excluding yourself from auto assignment

> There are some scenarios where you'd like to prevent yourself from being automatically assigned as a reviewer, e.g. holidays. To do this, simply click your avatar in the top right of GitHub, click "Set status" and then tick "Busy". Busy team members will not be automatically assigned as reviewers.

Reviewer 3: We should probably share this in `#ios-pedia`

Reviewer 2: Good point! It could be also updated [here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/OutOffOfficeRequest.md)
